File: internal/api/websocket.go
package api

import (
	"context"
	"encoding/base64"
	"encoding/json"
	"errors"
	"log/slog"
	"net"
	"net/http"
	"sync"
	"time"

	"github.com/coder/websocket"

	"github.com/nikicat/secrets-dispatcher/internal/approval"
	"github.com/nikicat/secrets-dispatcher/internal/proxy"
)

const (
	// Time allowed to write a message to the peer.
	writeWait = 10 * time.Second

	// Send pings to peer with this period.
	pingPeriod = 30 * time.Second

	// Maximum message size allowed from peer.
	maxMessageSize = 512
)

// WSMessage represents a message sent over the WebSocket.
type WSMessage struct {
	Type    string `json:"type"`
	Version string `json:"version,omitempty"`

	// For snapshot - no omitempty to ensure arrays are always present in JSON
	Requests                   []PendingRequest             `json:"requests"`
	Clients                    []proxy.ClientInfo           `json:"clients"`
	History                    []HistoryEntry               `json:"history"`
	AutoApproveRules           []approval.AutoApproveRule   `json:"auto_approve_rules"`
	SavedApprovalRules         []approval.SavedApprovalRule `json:"approval_rules"`
	TrustedSigners             []approval.TrustedSigner     `json:"trusted_signers"`
	TrustRules                 []approval.TrustRule         `json:"trust_rules"`
	AutoApproveDurationSeconds int                          `json:"auto_approve_duration_seconds,omitempty"`
	NotificationDelayMS        int                          `json:"notification_delay_ms,omitempty"`

	// For request_created
	Request *PendingRequest `json:"request,omitempty"`

	// For request_resolved
	ID     string `json:"id,omitempty"`
	Result string `json:"result,omitempty"`
	// Signature carries base64-encoded signature bytes for gpg_sign request_resolved events.
	// Empty for all other event types.
	Signature string `json:"signature,omitempty"`
	// GPGStatus carries the raw [GNUPG:] status lines from gpg --status-fd=2.
	// Written to thin client's stderr so git can parse gpg status.
	GPGStatus string `json:"gpg_status,omitempty"`
	// ExitCode carries gpg's exit code for failed signing attempts.
	// Non-zero only when real gpg exited with an error.
	ExitCode int `json:"exit_code,omitempty"`

	// For client_connected/client_disconnected
	Client *proxy.ClientInfo `json:"client,omitempty"`

	// For history_entry
	HistoryEntry *HistoryEntry `json:"history_entry,omitempty"`

	// For auto_approve_rule_added / auto_approve_rule_removed
	AutoApproveRule *approval.AutoApproveRule `json:"auto_approve_rule,omitempty"`
	// For approval_rule_added / approval_rule_updated / approval_rule_removed
	SavedApprovalRule *approval.SavedApprovalRule `json:"approval_rule,omitempty"`
}

// WSHandler handles WebSocket connections for real-time updates.
type WSHandler struct {
	manager             *approval.Manager
	clientProvider      ClientProvider
	auth                *Auth
	remoteSocket        string
	clientName          string
	notificationDelayMS int

	// Active connections
	connsMu sync.RWMutex
	conns   map[*wsConnection]struct{}
}

// NewWSHandler creates a new WebSocket handler.
func NewWSHandler(manager *approval.Manager, clientProvider ClientProvider, auth *Auth, remoteSocket, clientName string) *WSHandler {
	return &WSHandler{
		manager:        manager,
		clientProvider: clientProvider,
		auth:           auth,
		remoteSocket:   remoteSocket,
		clientName:     clientName,
		conns:          make(map[*wsConnection]struct{}),
	}
}

// wsConnection represents a single WebSocket connection.
type wsConnection struct {
	handler *WSHandler
	conn    *websocket.Conn
	send    chan []byte
	ctx     context.Context
	cancel  context.CancelFunc
}

// HandleWS handles WebSocket upgrade requests.
func (h *WSHandler) HandleWS(w http.ResponseWriter, r *http.Request) {
	// Validate session cookie or Bearer token before accepting upgrade.
	if !h.auth.ValidateRequest(r) {
		http.Error(w, `{"error": "unauthorized"}`, http.StatusUnauthorized)
		return
	}

	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
		// Allow same-origin connections only
		OriginPatterns: []string{"*"},
	})
	if err != nil {
		slog.Error("WebSocket accept failed", "error", err)
		return
	}

	conn.SetReadLimit(maxMessageSize)

	// Use background context - the WebSocket connection lives beyond the HTTP request
	ctx, cancel := context.WithCancel(context.Background())
	wsc := &wsConnection{
		handler: h,
		conn:    conn,
		send:    make(chan []byte, 256),
		ctx:     ctx,
		cancel:  cancel,
	}

	// Register connection
	h.connsMu.Lock()
	h.conns[wsc] = struct{}{}
	h.connsMu.Unlock()

	// Subscribe to manager events
	h.manager.Subscribe(wsc)

	// Send initial snapshot
	if err := wsc.sendSnapshot(); err != nil {
		slog.Error("Failed to send snapshot", "error", err)
		wsc.close()
		return
	}

	// Start reader and writer goroutines
	go wsc.writePump()
	go wsc.readPump()
}

// OnEvent implements approval.Observer.
func (wsc *wsConnection) OnEvent(event approval.Event) {
	var msgs []WSMessage

	switch event.Type {
	case approval.EventRequestCreated:
		msgs = append(msgs, WSMessage{
			Type:    "request_created",
			Request: convertRequest(event.Request),
		})
	case approval.EventRequestApproved:
		msg := WSMessage{
			Type:   "request_resolved",
			ID:     event.Request.ID,
			Result: "approved",
		}
		if event.Request.Type == approval.RequestTypeGPGSign {
			msg.Signature = base64.StdEncoding.EncodeToString(event.Request.Signature)
			msg.GPGStatus = string(event.Request.GPGStatus)
			if event.Request.GPGExitCode != 0 {
				msg.ExitCode = event.Request.GPGExitCode
			}
		}
		msgs = append(msgs, msg)
		entry := makeHistoryEntry(event.Request, "approved")
		msgs = append(msgs, WSMessage{
			Type:         "history_entry",
			HistoryEntry: &entry,
		})
	case approval.EventRequestDenied:
		msgs = append(msgs, WSMessage{
			Type:   "request_resolved",
			ID:     event.Request.ID,
			Result: "denied",
		})
		entry := makeHistoryEntry(event.Request, "denied")
		msgs = append(msgs, WSMessage{
			Type:         "history_entry",
			HistoryEntry: &entry,
		})
	case approval.EventRequestExpired:
		msgs = append(msgs, WSMessage{
			Type: "request_expired",
			ID:   event.Request.ID,
		})
		entry := makeHistoryEntry(event.Request, "expired")
		msgs = append(msgs, WSMessage{
			Type:         "history_entry",
			HistoryEntry: &entry,
		})
	case approval.EventRequestCancelled:
		msgs = append(msgs, WSMessage{
			Type: "request_cancelled",
			ID:   event.Request.ID,
		})
		entry := makeHistoryEntry(event.Request, "cancelled")
		msgs = append(msgs, WSMessage{
			Type:         "history_entry",
			HistoryEntry: &entry,
		})
	case approval.EventRequestAutoApproved:
		// GPG sign auto-approvals carry signature data that the thin client needs.
		if event.Request.Type == approval.RequestTypeGPGSign {
			msg := WSMessage{
				Type:   "request_resolved",
				ID:     event.Request.ID,
				Result: "auto_approved",
			}
			msg.Signature = base64.StdEncoding.EncodeToString(event.Request.Signature)
			msg.GPGStatus = string(event.Request.GPGStatus)
			if event.Request.GPGExitCode != 0 {
				msg.ExitCode = event.Request.GPGExitCode
			}
			msgs = append(msgs, msg)
		}
		entry := makeHistoryEntry(event.Request, "auto_approved")
		msgs = append(msgs, WSMessage{
			Type:         "history_entry",
			HistoryEntry: &entry,
		})
	case approval.EventRequestIgnored:
		entry := makeHistoryEntry(event.Request, "ignored")
		msgs = append(msgs, WSMessage{
			Type:         "history_entry",
			HistoryEntry: &entry,
		})
	case approval.EventAutoApproveRuleAdded:
		msgs = append(msgs, WSMessage{
			Type:            "auto_approve_rule_added",
			AutoApproveRule: event.Rule,
		})
	case approval.EventAutoApproveRuleRemoved:
		msgs = append(msgs, WSMessage{
			Type: "auto_approve_rule_removed",
			ID:   event.Rule.ID,
		})
	case approval.EventSavedApprovalRuleAdded:
		msgs = append(msgs, WSMessage{
			Type:              "approval_rule_added",
			SavedApprovalRule: event.SavedRule,
		})
	case approval.EventSavedApprovalRuleUpdated:
		msgs = append(msgs, WSMessage{
			Type:              "approval_rule_updated",
			SavedApprovalRule: event.SavedRule,
		})
	case approval.EventSavedApprovalRuleRemoved:
		msgs = append(msgs, WSMessage{
			Type: "approval_rule_removed",
			ID:   event.SavedRule.ID,
		})
	default:
		return
	}

	for _, msg := range msgs {
		data, err := json.Marshal(msg)
		if err != nil {
			slog.Error("Failed to marshal WebSocket message", "error", err)
			continue
		}

		// Non-blocking send - drop message if client is slow
		select {
		case wsc.send <- data:
			slog.Debug("ws send", "type", msg.Type, "id", msg.ID)
		default:
			slog.Warn("WebSocket send buffer full, dropping message", "type", msg.Type, "id", msg.ID)
		}
	}
}

// makeHistoryEntry creates a HistoryEntry for WebSocket messages.
func makeHistoryEntry(req *approval.Request, resolution string) HistoryEntry {
	items := make([]ItemInfo, len(req.Items))
	for i, item := range req.Items {
		items[i] = ItemInfo{
			Path:       item.Path,
			Label:      item.Label,
			Attributes: item.Attributes,
		}
	}
	return HistoryEntry{
		Request: PendingRequest{
			ID:               req.ID,
			Client:           req.Client,
			Items:            items,
			Session:          req.Session,
			CreatedAt:        req.CreatedAt,
			ExpiresAt:        req.ExpiresAt,
			Type:             string(req.Type),
			SearchAttributes: req.SearchAttributes,
			SenderInfo:       convertSenderInfo(req.SenderInfo),
			GPGSignInfo:      req.GPGSignInfo,
		},
		Resolution: resolution,
		ResolvedAt: time.Now(),
	}
}

// sendSnapshot sends the current state to the client.
func (wsc *wsConnection) sendSnapshot() error {
	h := wsc.handler

	// Get pending requests
	pending := h.manager.List()
	requests := make([]PendingRequest, len(pending))
	for i, req := range pending {
		requests[i] = *convertRequest(req)
	}

	// Get clients
	var clients []proxy.ClientInfo
	if h.clientProvider != nil {
		clients = h.clientProvider.Clients()
	} else {
		clients = []proxy.ClientInfo{
			{Name: h.clientName, SocketPath: h.remoteSocket},
		}
	}

	// Get history
	historyEntries := h.manager.History()
	history := make([]HistoryEntry, len(historyEntries))
	for i, entry := range historyEntries {
		history[i] = convertHistoryEntry(entry)
	}

	// Get auto-approve rules
	autoApproveRules := h.manager.ListAutoApproveRules()
	if autoApproveRules == nil {
		autoApproveRules = []approval.AutoApproveRule{}
	}
	savedApprovalRules := h.manager.ListSavedApprovalRules()
	if savedApprovalRules == nil {
		savedApprovalRules = []approval.SavedApprovalRule{}
	}

	// Get trusted signers
	trustedSigners := h.manager.ListTrustedSigners()
	if trustedSigners == nil {
		trustedSigners = []approval.TrustedSigner{}
	}

	// Get trust rules
	trustRules := h.manager.ListTrustRules()
	if trustRules == nil {
		trustRules = []approval.TrustRule{}
	}

	msg := WSMessage{
		Type:                       "snapshot",
		Version:                    BuildVersion,
		Requests:                   requests,
		Clients:                    clients,
		History:                    history,
		AutoApproveRules:           autoApproveRules,
		SavedApprovalRules:         savedApprovalRules,
		TrustedSigners:             trustedSigners,
		TrustRules:                 trustRules,
		AutoApproveDurationSeconds: int(h.manager.AutoApproveDuration().Seconds()),
		NotificationDelayMS:        h.notificationDelayMS,
	}

	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}

	// Send directly (not through channel) for initial snapshot
	ctx, cancel := context.WithTimeout(wsc.ctx, writeWait)
	defer cancel()
	slog.Debug("ws send snapshot", "requests", len(requests), "clients", len(clients), "history", len(history))
	return wsc.conn.Write(ctx, websocket.MessageText, data)
}

// writePump pumps messages from the send channel to the WebSocket connection.
func (wsc *wsConnection) writePump() {
	ticker := time.NewTicker(pingPeriod)
	defer func() {
		ticker.Stop()
		wsc.close()
	}()

	for {
		select {
		case <-wsc.ctx.Done():
			return

		case message, ok := <-wsc.send:
			if !ok {
				// Channel closed
				return
			}
			ctx, cancel := context.WithTimeout(wsc.ctx, writeWait)
			err := wsc.conn.Write(ctx, websocket.MessageText, message)
			cancel()
			if err != nil {
				if !isConnClosedErr(err) {
					slog.Debug("WebSocket write failed", "error", err)
				}
				return
			}

		case <-ticker.C:
			// Send ping
			ctx, cancel := context.WithTimeout(wsc.ctx, writeWait)
			err := wsc.conn.Ping(ctx)
			cancel()
			if err != nil {
				if !isConnClosedErr(err) {
					slog.Debug("WebSocket ping failed", "error", err)
				}
				return
			}
		}
	}
}

// readPump reads messages from the WebSocket connection.
// We don't expect any messages from the client, this is just for close detection.
func (wsc *wsConnection) readPump() {
	defer wsc.close()

	for {
		_, _, err := wsc.conn.Read(wsc.ctx)
		if err != nil {
			// Connection closed
			return
		}
		// Ignore any messages from client
	}
}

// close cleans up the connection.
func (wsc *wsConnection) close() {
	wsc.cancel()

	// Unsubscribe from manager
	wsc.handler.manager.Unsubscribe(wsc)

	// Unregister connection
	wsc.handler.connsMu.Lock()
	delete(wsc.handler.conns, wsc)
	wsc.handler.connsMu.Unlock()

	// Close the WebSocket
	wsc.conn.Close(websocket.StatusNormalClosure, "")
}

// convertRequest converts an approval.Request to an API PendingRequest.
func convertRequest(req *approval.Request) *PendingRequest {
	items := make([]ItemInfo, len(req.Items))
	for i, item := range req.Items {
		items[i] = ItemInfo{
			Path:       item.Path,
			Label:      item.Label,
			Attributes: item.Attributes,
		}
	}
	return &PendingRequest{
		ID:               req.ID,
		Client:           req.Client,
		Items:            items,
		Session:          req.Session,
		CreatedAt:        req.CreatedAt,
		ExpiresAt:        req.ExpiresAt,
		Type:             string(req.Type),
		SearchAttributes: req.SearchAttributes,
		SenderInfo:       convertSenderInfo(req.SenderInfo),
		GPGSignInfo:      req.GPGSignInfo,
	}
}

// isConnClosedErr returns true if the error indicates a connection that was
// already closed or broken (broken pipe, connection reset, context cancelled).
// These are expected during normal WebSocket teardown and not worth logging.
func isConnClosedErr(err error) bool {
	if errors.Is(err, context.Canceled) || errors.Is(err, net.ErrClosed) {
		return true
	}
	var netErr *net.OpError
	if errors.As(err, &netErr) {
		// Covers "write: broken pipe" and "write: connection reset by peer"
		return true
	}
	return false
}

// SetNotificationDelay sets the notification delay (in ms) sent to browser clients.
func (h *WSHandler) SetNotificationDelay(ms int) {
	h.notificationDelayMS = ms
}

// BroadcastClientConnected sends a client_connected message to all connections.
func (h *WSHandler) BroadcastClientConnected(client proxy.ClientInfo) {
	h.broadcast(WSMessage{
		Type:   "client_connected",
		Client: &client,
	})
}

// BroadcastClientDisconnected sends a client_disconnected message to all connections.
func (h *WSHandler) BroadcastClientDisconnected(client proxy.ClientInfo) {
	h.broadcast(WSMessage{
		Type:   "client_disconnected",
		Client: &client,
	})
}

// OnClientConnected implements proxy.ClientObserver.
func (h *WSHandler) OnClientConnected(client proxy.ClientInfo) {
	h.BroadcastClientConnected(client)
}

// OnClientDisconnected implements proxy.ClientObserver.
func (h *WSHandler) OnClientDisconnected(client proxy.ClientInfo) {
	h.BroadcastClientDisconnected(client)
}

// broadcast sends a message to all connected clients.
func (h *WSHandler) broadcast(msg WSMessage) {
	data, err := json.Marshal(msg)
	if err != nil {
		slog.Error("Failed to marshal broadcast message", "error", err)
		return
	}

	h.connsMu.RLock()
	defer h.connsMu.RUnlock()

	for wsc := range h.conns {
		select {
		case wsc.send <- data:
		default:
			// Drop if buffer full
		}
	}
}

File: web/src/lib/websocket.ts
import type {
  AutoApproveRule,
  ClientInfo,
  HistoryEntry,
  PendingRequest,
  SavedApprovalRule,
  TrustedSigner,
  TrustRule,
  WSMessage,
} from "./types";

export interface ApprovalWebSocketCallbacks {
  onSnapshot?: (
    requests: PendingRequest[],
    clients: ClientInfo[],
    history: HistoryEntry[],
    version: string,
    autoApproveRules: AutoApproveRule[],
    approvalRules: SavedApprovalRule[],
    trustedSigners: TrustedSigner[],
    trustRules: TrustRule[],
    autoApproveDurationSeconds: number,
    notificationDelayMS: number,
  ) => void;
  onRequestCreated?: (request: PendingRequest) => void;
  onRequestResolved?: (id: string, result: "approved" | "denied") => void;
  onRequestExpired?: (id: string) => void;
  onRequestCancelled?: (id: string) => void;
  onClientConnected?: (client: ClientInfo) => void;
  onClientDisconnected?: (client: ClientInfo) => void;
  onHistoryEntry?: (entry: HistoryEntry) => void;
  onAutoApproveRuleAdded?: (rule: AutoApproveRule) => void;
  onAutoApproveRuleRemoved?: (id: string) => void;
  onApprovalRuleAdded?: (rule: SavedApprovalRule) => void;
  onApprovalRuleUpdated?: (rule: SavedApprovalRule) => void;
  onApprovalRuleRemoved?: (id: string) => void;
  onConnectionChange?: (isConnected: boolean) => void;
  onAuthError?: () => void;
  onVersionMismatch?: () => void;
}

export class ApprovalWebSocket {
  private ws: WebSocket | null = null;
  private callbacks: ApprovalWebSocketCallbacks;
  private reconnectDelay = 1000;
  private maxReconnectDelay = 30000;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private shouldReconnect = true;
  private isConnected = false;
  private clientVersion: string | null = null;

  constructor(callbacks: ApprovalWebSocketCallbacks) {
    this.callbacks = callbacks;
  }

  connect(): void {
    if (this.ws) {
      return;
    }

    this.shouldReconnect = true;
    this.doConnect();
  }

  private doConnect(): void {
    const protocol = window.location.protocol === "https:" ? "wss:" : "ws:";
    const url = `${protocol}//${window.location.host}/api/v1/ws`;

    try {
      this.ws = new WebSocket(url);
    } catch {
      this.scheduleReconnect();
      return;
    }

    this.ws.onopen = () => {
      this.isConnected = true;
      this.reconnectDelay = 1000; // Reset on successful connection
      this.callbacks.onConnectionChange?.(true);
    };

    this.ws.onclose = (event) => {
      this.ws = null;
      const wasConnected = this.isConnected;
      this.isConnected = false;

      if (wasConnected) {
        this.callbacks.onConnectionChange?.(false);
      }

      // WebSocket close code 1008 is explicit policy violation (auth error)
      if (event.code === 1008) {
        this.callbacks.onAuthError?.();
        this.shouldReconnect = false;
      }

      // For code 1006 (abnormal closure) when we never connected,
      // verify auth status before assuming it's an auth error
      if (event.code === 1006 && !wasConnected) {
        this.verifyAuthAndReconnect();
        return;
      }

      if (this.shouldReconnect) {
        this.scheduleReconnect();
      }
    };

    this.ws.onerror = () => {
      // Error handling is done in onclose
    };

    this.ws.onmessage = (event) => {
      this.handleMessage(event.data);
    };
  }

  disconnect(): void {
    this.shouldReconnect = false;

    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }

    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }

    if (this.isConnected) {
      this.isConnected = false;
      this.callbacks.onConnectionChange?.(false);
    }
  }

  private handleMessage(data: string): void {
    let msg: WSMessage;
    try {
      msg = JSON.parse(data);
    } catch {
      console.error("Failed to parse WebSocket message:", data);
      return;
    }

    switch (msg.type) {
      case "snapshot":
        // Check for version mismatch on reconnect
        if (this.clientVersion === null) {
          // First connection - store version
          this.clientVersion = msg.version ?? "";
        } else if (this.clientVersion !== (msg.version ?? "")) {
          // Version mismatch - server was updated
          this.callbacks.onVersionMismatch?.();
          return;
        }
        this.callbacks.onSnapshot?.(
          msg.requests ?? [],
          msg.clients ?? [],
          msg.history ?? [],
          msg.version ?? "",
          msg.auto_approve_rules ?? [],
          msg.approval_rules ?? [],
          msg.trusted_signers ?? [],
          msg.trust_rules ?? [],
          msg.auto_approve_duration_seconds ?? 120,
          msg.notification_delay_ms ?? 0,
        );
        break;
      case "request_created":
        this.callbacks.onRequestCreated?.(msg.request);
        break;
      case "request_resolved":
        this.callbacks.onRequestResolved?.(msg.id, msg.result);
        break;
      case "request_expired":
        this.callbacks.onRequestExpired?.(msg.id);
        break;
      case "request_cancelled":
        this.callbacks.onRequestCancelled?.(msg.id);
        break;
      case "client_connected":
        this.callbacks.onClientConnected?.(msg.client);
        break;
      case "client_disconnected":
        this.callbacks.onClientDisconnected?.(msg.client);
        break;
      case "history_entry":
        this.callbacks.onHistoryEntry?.(msg.history_entry);
        break;
      case "auto_approve_rule_added":
        this.callbacks.onAutoApproveRuleAdded?.(msg.auto_approve_rule);
        break;
      case "auto_approve_rule_removed":
        this.callbacks.onAutoApproveRuleRemoved?.(msg.id);
        break;
      case "approval_rule_added":
        this.callbacks.onApprovalRuleAdded?.(msg.approval_rule);
        break;
      case "approval_rule_updated":
        this.callbacks.onApprovalRuleUpdated?.(msg.approval_rule);
        break;
      case "approval_rule_removed":
        this.callbacks.onApprovalRuleRemoved?.(msg.id);
        break;
      case "ping":
        // Server ping, no action needed
        break;
    }
  }

  private async verifyAuthAndReconnect(): Promise<void> {
    // Check if the server is up and we're still authenticated
    try {
      const response = await fetch("/api/v1/status", {
        credentials: "include",
      });

      if (response.status === 401) {
        // Confirmed auth error
        this.callbacks.onAuthError?.();
        this.shouldReconnect = false;
        return;
      }

      // Server is up and we're authenticated (or server is up but returned other error)
      // Continue reconnecting
      if (this.shouldReconnect) {
        this.scheduleReconnect();
      }
    } catch {
      // Network error - server probably not up yet, continue reconnecting
      if (this.shouldReconnect) {
        this.scheduleReconnect();
      }
    }
  }

  private scheduleReconnect(): void {
    if (this.reconnectTimer) {
      return;
    }

    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      if (this.shouldReconnect) {
        this.doConnect();
      }
    }, this.reconnectDelay);

    // Exponential backoff
    this.reconnectDelay = Math.min(
      this.reconnectDelay * 2,
      this.maxReconnectDelay,
    );
  }

  get connected(): boolean {
    return this.isConnected;
  }
}
